### PR TITLE
Regenerate project with cmake-init

### PR DESCRIPTION
**Pull request by friendlyanon**
This makes the project use modern CMake, spellchecker and clang-format linting in CI, vcpkg for dependencies and cleans up the tests.

Note that while vcpkg does have a port for nanobench, it does not properly install a CMake package, so an overlay port had to be written for that.  
Similarly, I created overlay ports for the other daw library dependencies, because I didn't want to wait for PRs to merge on those (also lazy). If you wish to make the overlay port as minimal as possible, you can use [cmake-init](https://github.com/friendlyanon/cmake-init) to generate a project structure for those.

For the tests, a structure similar to that of CMake's test suite was used by having variables affect the tests rather than arguments, which makes passing the same set of arguments to tests cleaner.

The content section has been removed from the README, because GitHub already implements a Table of Contents feature for READMEs, so this serves no purpose in the document itself.

Travis CI script has been nuked as they no longer provide CI services.

The GitHub Actions workflow has been updated with a code coverage job. Follow the instructions in the comments to enable that job.

---
#### Nuke `docs` from version control

You can automatically deploy documentation with every successful CI build. Doxygen generated docs just bloat the repository size unnecessarily. cmake-init generates projects with such a CI job, [here](https://github.com/friendlyanon/cmake-init-shared-static/blob/master/.github/workflows/ci.yml#L125) is one working example.

#### Nuke `test_data` from version control

Or at least have it on an orphaned branch that is pulled at configure time. Same reason as above, just bloats the repository.

#### Use the build system for the `DAW_JSON_VER` define

If I remember correctly from your CppCast session, you don't intend to keep ABI stability, so you can easily do this:

```cmake
target_compile_definitions(daw_json_link INTERFACE "DAW_JSON_VER=\
v${PROJECT_VERSION_MAJOR}_${PROJECT_VERSION_MINOR}_${PROJECT_VERSION_PATCH}")
```

and now you have a ~~single~~ less places to change the version number: The lists file and vcpkg.json.

---

Fixes: #238

### CODE REVIEW

**Comment by friendlyanon** [comment]
Since I don't see it running here, I kicked off a CI run on my fork: https://github.com/friendlyanon/daw_json_link/actions/runs/946487217

**Comment by beached** [comment]
This will take me a big to go through.  Couple of start notes:
* The ToC needs to stay as it was, it was not all content from the Readme but included links to other documentation files, not all views of it are on github, and I prefer to have everything up front.  
* I want to change the find_package of the deps from required to optional and if they are not there use fetchcontent. Or at least have an option to switch between. It's much easier for rapid updates during development.
* the clang-format on commit has been tried in the past and ended up blocking when files had been formatted.  Also it allows the commit for formatting updates, if needed, to be done without blocking CI.
* I'll have to try the spellcheck, I already some some exclusions and find the spell checkers in IDE's are often not recognizing terms of art
* I will look at the rest this weekend :)

**Comment by friendlyanon** [comment]
I restored the TOC and removed clang-format from the first job. Differences in CI and local clang-format outputs appear when the versions are mistmatched. If you use 11 in CI and 11 locally, then CI will not fail.

To use FetchContent instead of picking dependencies up from an install location, there are many ways of going about it, but I think this is a pretty efficient way of going about it:

* Create a file with the FetchContent commands you want and commands to stub out `find_package` calls:
```cmake
# Stub out dependencies provided with FetchContent
foreach(dep IN ITEMS daw_header_libraries daw_utf_range)
  # Empty script, so CMake does nothing
  file(WRITE "${PROJECT_BINARY_DIR}/stubs/${dep}Config.cmake" "")
  set("${dep}_DIR" "${PROJECT_BINARY_DIR}/stubs" CACHE PATH "")
endforeach()

include(FetchContent)

# ...
```
* Inject the file with [`CMAKE_PROJECT_daw_json_link_INCLUDE`](https://cmake.org/cmake/help/latest/command/project.html#code-injection), preferably by just setting this in a user preset. Refer to HACKING document for this.

If you want to test how things are in CI, then you can just delete the cache variables above and let CMake find the installed libraries instead. Or have a separate user preset that you use just to see if the project configures.

**Comment by friendlyanon** [comment]
Resolved conflicts, ran CI for the new commits and [everything passes](https://github.com/friendlyanon/daw_json_link/actions/runs/966311731)!

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+          env: '{ "CXX": "g++-10" }'
+          flags: -D DAW_NO_FLATTEN=YES
+
+    runs-on: ${{ matrix.os }}-latest
+
+    steps:
+    - uses: actions/checkout@v1
+
+    - name: Install vcpkg
+      uses: lukka/run-vcpkg@v7
```
Comment (beached) [inline .github/workflows/ci.yml]: This is MIT licensed and incompatible with BSL.  Still working my way down, but I don't understand what this does.  The two dependencies are super small and header only.

File: cookbook/enums.md [hunk 2]
```diff
@@ -11,43 +11,43 @@ There is no direct support for enum's in JSON.  However, they are often encoded
 The above JSON object has a member, `"member0"`, that is an array of strings.
 In order to store the enum as a JSON string, an overload for `to_string` and `from_string` are required. `to_string` is required to return a `Container` where the `begin( )` iterator's `value_type` is `char, not necessarily a `std::string`.
 
-To see a working example using this code, refer to [cookbook_enums1_test.cpp](../tests/src/cookbook_enums1_test.cpp) 
+To see a working example using this code, refer to [cookbook_enums1_test.cpp](../tests/src/cookbook_enums1_test.cpp)
 ```c++
-enum class Colours : uint8_t { red, green, blue, black };
```
Comment (beached) [inline cookbook/enums.md]: Not all of us are American, not sure I am liking the spellchecking.  I can do that in IDE(apparently I haven't been in a couple of the files.)

File: ports/daw-header-libraries/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,23 @@
+set(VCPKG_BUILD_TYPE release)
+
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO beached/header_libraries
+    REF 760902a78b23ebc79c2ac51384d081df8bc8022f
+    SHA512 36d36d5cfe591b179ba9bc0ab3da0a035ddefa5c97a8b56e80c9b80e216fd36fb0e8c94a60d8861efdc2bbb29370c9bad796bd527e0d44afe0f8eaf5a9f2816a
```
Comment (beached) [inline ports/daw-header-libraries/portfile.cmake]: Just so I understands, this means that one has to update these prior to any push to release?

File: readme.md [hunk 4]
```diff
@@ -281,16 +239,16 @@ using iterator_t = daw::json::json_array_iterator_trusted<MyClass>;
...
-If you want to serialize to JSON 
+If you want to serialize to JSON
 
 ```c++
 std::string my_json_data = to_json( MyClass{} );
 ```
 
 Or serialize a collection of things
 ```c++
-std::vector<MyClass> arry = ...;
```
Comment (beached) [inline readme.md]: I want to keep arry, it was on purpose

File: tests/src/daw_json_link_test.cpp [hunk 5]
```diff
@@ -32,7 +32,7 @@
 Temporarily disable the constexpr tests in MSVC when C++20
 */
 #if defined( _MSC_VER ) and not defined( __clang__ ) and \
-  defined( DAW_CXX_STANDARD ) and DAW_CXX_STANDARD == 20
+  __cplusplus > 201703L
```
Comment (beached) [inline tests/src/daw_json_link_test.cpp:35]: This breaks on MSVC as the version is locked at the 98 release.  These all need to be buildable without cmake/vcpkg.  This is working around a compiler bug in MSVC when /std:c++latest is enabled which enables their C++20 code paths.

**Comment by beached** [comment]
I don't understand.  So in the previous, fmtlib/nanobench where brought in only if testing was enabled.  But I don't see how that happens anymore.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (friendlyanon) [reply .github/workflows/ci.yml]: What does `this` refer to here? vcpkg? The `run-vcpkg` action? You were already using nanobench, which is also MIT license. License incompatibilities only matter for distribution and runtime usage generally.

The purpose of this action is to set vcpkg and the necessary env variables up at the given commit conveniently.

File: cookbook/enums.md (hunk 2 above)
Comment (friendlyanon) [reply cookbook/enums.md]: I'm not from the USA either, it just felt more natural in a programming context to not use British English.

File: readme.md (hunk 4 above)
Comment (friendlyanon) [reply readme.md]: OK, I can add the lines to be ignored.

File: tests/src/daw_json_link_test.cpp (hunk 5 above)
Comment (friendlyanon) [reply tests/src/daw_json_link_test.cpp:35]: https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus

This flag is passed in the `flags-windows` presets.

**Comment by friendlyanon** [comment]
There is a [`test` feature](https://github.com/friendlyanon/daw_json_link/blob/cmake/vcpkg.json#L10) which is enabled by the [`dev-mode` preset](https://github.com/friendlyanon/daw_json_link/blob/cmake/CMakePresets.json#L22).

File: ports/daw-header-libraries/portfile.cmake (hunk 3 above)
Comment (friendlyanon) [reply ports/daw-header-libraries/portfile.cmake]: Yes. If the semver version doesn't change, but the port does change, then the [port version](https://github.com/microsoft/vcpkg/blob/master/docs/users/versioning.md#port-version) has to be updated.

File: cookbook/enums.md (hunk 2 above)
Comment (friendlyanon) [reply cookbook/enums.md]: This change was implemented.

File: readme.md (hunk 4 above)
Comment (friendlyanon) [reply readme.md]: This change was implemented.

**Comment by friendlyanon** [comment]
Resolved conflicts.

**Comment by friendlyanon** [comment]
Resolved conflicts.

**Comment by beached** [comment]
Cherry picked a bunch of items and really appreciate the work.  I am not ready to adjust my workflow at this time for the other changes though.

**Pull request closed without merging**